src/main/java/org/seqdoop/hadoop_bam/BAMInputFormat.java

```diff
@@ -308,60 +303,79 @@ private int addProbabilisticSplits(
 					.setUseAsyncIo(false);
 			SamReader samReader = readerFactory.open(NIOFileUtil.asPath(fs.makeQualified(bamFile).toUri()));
 			if (!samReader.hasIndex()) {
-				System.err.println("WARNING: no BAM index file found, splits will not be " +
-						"filtered, which may be very inefficient: " + bamFile);
-				return splits;
+				throw new IllegalArgumentException("Intervals set but no BAM index file found for " + bamFile);
 			}
 
 			FSDataInputStream in = fs.open(bamFile);
 			SAMFileHeader header = SAMHeaderReader.readSAMHeaderFrom(in, conf);
 			SAMSequenceDictionary dict = header.getSequenceDictionary();
 			BAMIndex idx = samReader.indexing().getIndex();
-
-			for (Locatable interval : intervals) {
-				int referenceIndex = dict.getSequenceIndex(interval.getContig());
-				int intervalStart = interval.getStart();
-				int intervalEnd = interval.getEnd();
-				BAMFileSpan sfs = idx.getSpanOverlapping(referenceIndex, intervalStart, intervalEnd);
-				chunks.addAll(sfs.getChunks());
-			}
+			QueryInterval[] queryIntervals = prepareQueryIntervals(intervals, dict);
+			fileToSpan.put(bamFile, BAMFileReader.getFileSpan(queryIntervals, idx));
```

Comment:
And also acquire the stream inside of a try with resources, since I think almost all of these methods can wind up throwing.